File: dmn-test-cases/standard/tck/1.3/cl3/0085-decision-services/translator/expected/java/dmn/DecisionService_011.java

import java.util.*;
import java.util.stream.Collectors;

@javax.annotation.Generated(value = {"ds.ftl", "decisionService_011"})
@com.gs.dmn.runtime.annotation.DRGElement(
    namespace = "",
    name = "decisionService_011",
    label = "",
    elementKind = com.gs.dmn.runtime.annotation.DRGElementKind.DECISION_SERVICE,
    expressionKind = com.gs.dmn.runtime.annotation.ExpressionKind.OTHER,
    hitPolicy = com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
    rulesCount = -1
)
public class DecisionService_011 extends com.gs.dmn.runtime.JavaTimeDMNBaseDecision<String> {
    public static final com.gs.dmn.runtime.listener.DRGElement DRG_ELEMENT_METADATA = new com.gs.dmn.runtime.listener.DRGElement(
        "",
        "decisionService_011",
        "",
        com.gs.dmn.runtime.annotation.DRGElementKind.DECISION_SERVICE,
        com.gs.dmn.runtime.annotation.ExpressionKind.OTHER,
        com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
        -1
    );

    private static class DecisionService_011LazyHolder {
        static final DecisionService_011 INSTANCE = new DecisionService_011();
    }
    public static DecisionService_011 instance() {
        return DecisionService_011LazyHolder.INSTANCE;
    }

    private final Decision_011_2 decision_011_2;

    private DecisionService_011() {
        this(new Decision_011_2());
    }

    private DecisionService_011(Decision_011_2 decision_011_2) {
        this.decision_011_2 = decision_011_2;
    }

    @java.lang.Override()
    public String applyMap(java.util.Map<String, String> input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return apply(input_.get("inputData_011_1"), input_.get("inputData_011_2"), input_.get("decision_011_3"), input_.get("decision_011_4"), context_);
        } catch (Exception e) {
            logError("Cannot apply decision 'DecisionService_011'", e);
            return null;
        }
    }

    public String apply(String inputData_011_1, String inputData_011_2, String decision_011_3, String decision_011_4, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            // Start DS 'decisionService_011'
            com.gs.dmn.runtime.annotation.AnnotationSet annotationSet_ = context_ != null ? context_.getAnnotations() : null;
            com.gs.dmn.runtime.listener.EventListener eventListener_ = context_ != null ? context_.getEventListener() : null;
            com.gs.dmn.runtime.external.ExternalFunctionExecutor externalExecutor_ = context_ != null ? context_.getExternalFunctionExecutor() : null;
            com.gs.dmn.runtime.cache.Cache cache_ = context_ != null ? context_.getCache() : null;
            long decisionService_011StartTime_ = System.currentTimeMillis();
            com.gs.dmn.runtime.listener.Arguments decisionService_011Arguments_ = new com.gs.dmn.runtime.listener.Arguments();
            decisionService_011Arguments_.put("inputData_011_1", inputData_011_1);
            decisionService_011Arguments_.put("inputData_011_2", inputData_011_2);
            decisionService_011Arguments_.put("decision_011_3", decision_011_3);
            decisionService_011Arguments_.put("decision_011_4", decision_011_4);
            eventListener_.startDRGElement(DRG_ELEMENT_METADATA, decisionService_011Arguments_);

            // Bind input decisions
            cache_.bind("decision_011_3", decision_011_3);
            cache_.bind("decision_011_4", decision_011_4);

            // Evaluate DS 'decisionService_011'
            String output_ = lambda.apply(inputData_011_1, inputData_011_2, decision_011_3, decision_011_4, context_);

            // End DS 'decisionService_011'
            eventListener_.endDRGElement(DRG_ELEMENT_METADATA, decisionService_011Arguments_, output_, (System.currentTimeMillis() - decisionService_011StartTime_));

            return output_;
        } catch (Exception e) {
            logError("Exception caught in 'decisionService_011' evaluation", e);
            return null;
        }
    }

    public com.gs.dmn.runtime.LambdaExpression<String> lambda =
        new com.gs.dmn.runtime.LambdaExpression<String>() {
            public String apply(Object... args_) {
                String inputData_011_1 = 0 < args_.length ? (String) args_[0] : null;
                String inputData_011_2 = 1 < args_.length ? (String) args_[1] : null;
                String decision_011_3 = 2 < args_.length ? (String) args_[2] : null;
                String decision_011_4 = 3 < args_.length ? (String) args_[3] : null;
                com.gs.dmn.runtime.ExecutionContext context_ = 4 < args_.length ? (com.gs.dmn.runtime.ExecutionContext) args_[4] : null;
                com.gs.dmn.runtime.annotation.AnnotationSet annotationSet_ = context_ != null ? context_.getAnnotations() : null;
                com.gs.dmn.runtime.listener.EventListener eventListener_ = context_ != null ? context_.getEventListener() : null;
                com.gs.dmn.runtime.external.ExternalFunctionExecutor externalExecutor_ = context_ != null ? context_.getExternalFunctionExecutor() : null;
                com.gs.dmn.runtime.cache.Cache cache_ = context_ != null ? context_.getCache() : null;

                // Apply child decisions
                String decision_011_2 = DecisionService_011.this.decision_011_2.apply(inputData_011_1, inputData_011_2, context_);

                return decision_011_2;
            }
    };
}

File: dmn-test-cases/standard/tck/1.3/cl3/0085-decision-services/translator/expected/python/dmn/DecisionService_011.py
import typing
import decimal
import datetime
import isodate
import time

import jdmn.runtime.Context
import jdmn.runtime.DefaultDMNBaseDecision
import jdmn.runtime.ExecutionContext
import jdmn.runtime.LambdaExpression
import jdmn.runtime.LazyEval
import jdmn.runtime.Pair
import jdmn.runtime.Range
import jdmn.runtime.RuleOutput
import jdmn.runtime.RuleOutputList

import jdmn.runtime.annotation.Annotation
import jdmn.runtime.annotation.AnnotationSet
import jdmn.runtime.annotation.DRGElementKind
import jdmn.runtime.annotation.ExpressionKind
import jdmn.runtime.annotation.HitPolicy

import jdmn.runtime.cache.Cache

import jdmn.runtime.external.ExternalFunctionExecutor

import jdmn.runtime.listener.Arguments
import jdmn.runtime.listener.DRGElement
import jdmn.runtime.listener.EventListener
import jdmn.runtime.listener.Rule

import Decision_011_2


# Generated(value = ["ds.ftl", "decisionService_011"])
class DecisionService_011(jdmn.runtime.DefaultDMNBaseDecision.DefaultDMNBaseDecision):
    DRG_ELEMENT_METADATA: jdmn.runtime.listener.DRGElement.DRGElement = jdmn.runtime.listener.DRGElement.DRGElement(
        "",
        "decisionService_011",
        "",
        jdmn.runtime.annotation.DRGElementKind.DRGElementKind.DECISION_SERVICE,
        jdmn.runtime.annotation.ExpressionKind.ExpressionKind.OTHER,
        jdmn.runtime.annotation.HitPolicy.HitPolicy.UNKNOWN,
        -1
    )
    _instance = None

    def __init__(self):
        raise RuntimeError("Call instance() instead")

    @classmethod
    def instance(cls):
        if cls._instance is None:
            cls._instance = cls.__new__(cls)
            jdmn.runtime.DefaultDMNBaseDecision.DefaultDMNBaseDecision.__init__(cls._instance)
            cls._instance.initSubDecisions()
        return cls._instance

    def initSubDecisions(self, decision_011_2: Decision_011_2.Decision_011_2 = None):
        self.decision_011_2 = Decision_011_2.Decision_011_2() if decision_011_2 is None else decision_011_2

    def apply(self, inputData_011_1: typing.Optional[str], inputData_011_2: typing.Optional[str], decision_011_3: typing.Optional[str], decision_011_4: typing.Optional[str], annotationSet_: jdmn.runtime.annotation.AnnotationSet.AnnotationSet, eventListener_: jdmn.runtime.listener.EventListener.EventListener, externalExecutor_: jdmn.runtime.external.ExternalFunctionExecutor.ExternalFunctionExecutor, cache_: jdmn.runtime.cache.Cache.Cache) -> typing.Optional[str]:
        try:
            # Start DS 'decisionService_011'
            decisionService_011StartTime_ = int(time.time_ns()/1000)
            decisionService_011Arguments_ = jdmn.runtime.listener.Arguments.Arguments()
            decisionService_011Arguments_.put("inputData_011_1", inputData_011_1)
            decisionService_011Arguments_.put("inputData_011_2", inputData_011_2)
            decisionService_011Arguments_.put("decision_011_3", decision_011_3)
            decisionService_011Arguments_.put("decision_011_4", decision_011_4)
            eventListener_.startDRGElement(self.DRG_ELEMENT_METADATA, decisionService_011Arguments_)

            # Bind input decisions
            cache_.bind("decision_011_3", decision_011_3)
            cache_.bind("decision_011_4", decision_011_4)

            # Evaluate DS 'decisionService_011'
            output_: typing.Optional[str] = self.evaluate(inputData_011_1, inputData_011_2, decision_011_3, decision_011_4, annotationSet_, eventListener_, externalExecutor_, cache_)

            # End DS 'decisionService_011'
            eventListener_.endDRGElement(self.DRG_ELEMENT_METADATA, decisionService_011Arguments_, output_, (int(time.time_ns()/1000) - decisionService_011StartTime_))

            return output_
        except Exception as e:
            self.logError("Exception caught in 'decisionService_011' evaluation", e)
            return None

    def evaluate(self, inputData_011_1: typing.Optional[str], inputData_011_2: typing.Optional[str], decision_011_3: typing.Optional[str], decision_011_4: typing.Optional[str], annotationSet_: jdmn.runtime.annotation.AnnotationSet.AnnotationSet, eventListener_: jdmn.runtime.listener.EventListener.EventListener, externalExecutor_: jdmn.runtime.external.ExternalFunctionExecutor.ExternalFunctionExecutor, cache_: jdmn.runtime.cache.Cache.Cache) -> typing.Optional[str]:
        # Apply child decisions
        decision_011_2: typing.Optional[str] = self.decision_011_2.apply(inputData_011_1, inputData_011_2, annotationSet_, eventListener_, externalExecutor_, cache_)

        return decision_011_2
